core/slice.go

```diff
@@ -316,6 +318,15 @@ func (sl *Slice) Append(header *types.Header, domPendingHeader *types.Header, do
 		time9 = common.PrettyDuration(time.Since(start))
 
 	}
+
+	// If block hash at the beginning of the append is not the same as before
+	// modifying the phCache and Batch write we need to reject the block. This
+	// check can be removed after fixing the block hash mutation bug observed in
+	// testnet
+	if block.Hash() != blockHash {
+		return nil, false, false, ErrBlockHashChanged
```

Comment:
What about other side effects like SetCurrentState(), do we have to undo those?

Reply:
it wont cause any side effect, setting current state is done so that we can recalculate the header root and verify the state transition, so running SetCurrentState is a fine expense